File: references/twitter/session.py
#!/usr/bin/env python3
"""Persistent camoufox profile holding the x.com login.

Scraping cookies out of Chrome is fragile — it depends on the keychain, on
guessing which of N Chrome profiles is signed in, and on the user happening to
be logged in there at all. A persistent camoufox profile removes all of that:
the user signs in once through `--login`, x.com issues its cookies to that
profile, and every later run reuses them headlessly.

We never see or store a password. Only the cookies x.com hands out live on
disk, in the profile directory.
"""

import os
import time
from pathlib import Path

from lib.tui import OK, STEP, WARN

from twitter.browser_launch import launch_camoufox_persistent
from twitter.cookies import SESSION_COOKIE_NAME

PROFILE_DIR = Path(
    os.environ.get("TWITTER_PROFILE_DIR", "~/.twitter-camoufox-profile")
).expanduser()

LOGIN_URL = os.environ.get("TWITTER_LOGIN_URL", "https://x.com/i/flow/login")
LOGIN_TIMEOUT_S = float(os.environ.get("TWITTER_LOGIN_TIMEOUT_S", "300"))
LOGIN_POLL_S = float(os.environ.get("TWITTER_LOGIN_POLL_S", "2"))
# Firefox stores a populated profile as a directory tree; an empty dir means
# `--login` was never completed.
_PROFILE_MARKER = "cookies.sqlite"


def profile_exists() -> bool:
    """True when a camoufox profile directory has actually been populated."""
    return (PROFILE_DIR / _PROFILE_MARKER).exists()


def context_has_session(context) -> bool:
    """True when the live browser context holds an x.com session token."""
    try:
        cookies = context.cookies()
    except Exception:
        return False
    return any(
        c.get("name") == SESSION_COOKIE_NAME and c.get("value") for c in cookies
    )


def open_session(headless: bool = True):
    """Open the persistent profile. Returns (context, closer)."""
    PROFILE_DIR.mkdir(parents=True, exist_ok=True)
    return launch_camoufox_persistent(PROFILE_DIR, headless=headless)


def has_saved_session() -> bool:
    """True when the on-disk profile holds an x.com session token.

    Reads the profile's own cookies.sqlite rather than launching a browser, so
    the caller can choose a backend without paying a browser start-up.
    """
    if not profile_exists():
        return False
    import shutil
    import sqlite3
    import tempfile

    tf = tempfile.NamedTemporaryFile(suffix=".sqlite", delete=False)
    tmp = Path(tf.name)
    tf.close()
    try:
        shutil.copy2(PROFILE_DIR / _PROFILE_MARKER, tmp)
        conn = sqlite3.connect(f"file:{tmp}?mode=ro", uri=True)
        row = conn.execute(
            "SELECT COUNT(*) FROM moz_cookies "
            "WHERE name = ? AND value != '' AND (host = '.x.com' OR host = 'x.com')",
            (SESSION_COOKIE_NAME,),
        ).fetchone()
        conn.close()
        return bool(row and row[0])
    except Exception:
        return False
    finally:
        tmp.unlink(missing_ok=True)


def login(timeout_s: float = LOGIN_TIMEOUT_S) -> bool:
    """Open a real window so the user can sign in to x.com themselves.

    Polls until x.com issues a session cookie to the profile. Returns True on
    success. The password is typed by the user into the browser — it never
    passes through this process.
    """
    print(f"{STEP} Opening x.com in a camoufox window — sign in there.")
    print(f"{STEP} Profile: {PROFILE_DIR}")
    print(f"{STEP} Waiting up to {timeout_s:.0f}s for the session cookie ...")

    context, closer = open_session(headless=False)
    try:
        page = context.pages[0] if context.pages else context.new_page()
        try:
            page.goto(LOGIN_URL, wait_until="domcontentloaded", timeout=60_000)
        except Exception as e:
            print(f"{WARN} Could not open {LOGIN_URL}: {e}")

        deadline = time.monotonic() + timeout_s
        while time.monotonic() < deadline:
            if context_has_session(context):
                print(f"{OK} Signed in — session saved to {PROFILE_DIR}")
                print(f"{OK} Future runs are headless; no login needed.")
                return True
            time.sleep(LOGIN_POLL_S)

        print(f"{WARN} Timed out after {timeout_s:.0f}s without a session cookie.")
        print(f"{WARN} Re-run the login and complete the sign-in in the window.")
        return False
    finally:
        # Closing the context is what flushes cookies.sqlite to disk.
        closer()

File: rust/src/ztools/twitter/session.rs
//! Persistent camoufox profile holding the x.com login.
//!
//! Port of `twitter/session.py`: the user signs in once through a headed
//! `--login` run, x.com issues its cookies to the profile directory, and every
//! later headless run reuses them. No password is ever seen or stored — only
//! the cookies x.com hands out, on disk, in the profile directory.

use std::path::{Path, PathBuf};
use std::time::{Duration, Instant};

/// Profile location. `TWITTER_PROFILE_DIR` overrides the default, exactly like
/// the Python `PROFILE_DIR`.
#[must_use]
pub fn profile_dir() -> PathBuf {
    if let Ok(raw) = std::env::var("TWITTER_PROFILE_DIR") {
        return PathBuf::from(raw);
    }
    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("/root"))
        .join(".twitter-camoufox-profile")
}

/// Login page and polling bounds from `session.py`.
pub const LOGIN_URL: &str = "https://x.com/i/flow/login";
pub const LOGIN_TIMEOUT: Duration = Duration::from_secs(300);
pub const LOGIN_POLL: Duration = Duration::from_secs(2);
/// A populated Firefox profile always contains this file; an empty dir means
/// `--login` was never completed.
pub const PROFILE_MARKER: &str = "cookies.sqlite";

/// True when the profile directory was actually populated by a login run.
#[must_use]
pub fn profile_exists(dir: &Path) -> bool {
    dir.join(PROFILE_MARKER).is_file()
}

/// True when the on-disk profile holds an x.com session token.
///
/// Reads the profile's own `cookies.sqlite` through the same Firefox-family
/// reader the installed-browser path uses — no browser launch, so this is
/// cheap enough to check before every headless run.
///
/// # Errors
///
/// Only when the marker file is missing (login never completed). An unreadable
/// but present DB reads as no-session, never as an error: a corrupt profile
/// and a logged-out profile demand the same remedy (`--login`).
pub fn has_saved_session(dir: &Path) -> Result<bool, SavedSessionError> {
    if !profile_exists(dir) {
        return Err(SavedSessionError::NoProfile);
    }
    let db = dir.join(PROFILE_MARKER);
    let cookies = super::cookies::read_firefox_cookies(&db, super::cookies::DEFAULT_DOMAINS)
        .unwrap_or_default();
    Ok(super::cookies::has_session_cookie(&cookies))
}

/// Why a saved session cannot be used.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SavedSessionError {
    /// The profile directory was never populated — run `--login` first.
    NoProfile,
}

impl std::fmt::Display for SavedSessionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NoProfile => write!(
                f,
                "no saved x.com session — run `twitter --login` once, then retry"
            ),
        }
    }
}

impl std::error::Error for SavedSessionError {}

/// Poll `has_session` until it reports true or `timeout` elapses.
///
/// Pure over injected time + probe so the login wait is testable without a
/// browser: the deadline, the poll interval, and the "user closed the window"
/// abort all surface here, not in the driver.
///
/// # Errors
///
/// [`WaitError::Aborted`] when the login window goes away first;
/// [`WaitError::TimedOut`] when no session appears before `timeout`.
pub fn wait_for_session(
    mut has_session: impl FnMut() -> bool,
    mut aborted: impl FnMut() -> bool,
    timeout: Duration,
    poll: Duration,
    mut now: impl FnMut() -> Instant,
    mut sleep: impl FnMut(Duration),
) -> Result<(), WaitError> {
    let start = now();
    loop {
        if has_session() {
            return Ok(());
        }
        if aborted() {
            return Err(WaitError::Aborted);
        }
        if now().duration_since(start) >= timeout {
            return Err(WaitError::TimedOut);
        }
        sleep(poll);
    }
}

/// Why a headed login wait ended without a session.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum WaitError {
    /// The login window went away before any session appeared.
    Aborted,
    /// `LOGIN_TIMEOUT` elapsed with no session cookie.
    TimedOut,
}

impl std::fmt::Display for WaitError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Aborted => write!(f, "login window closed before sign-in completed"),
            Self::TimedOut => write!(f, "no x.com session appeared before the login timeout"),
        }
    }
}

impl std::error::Error for WaitError {}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cell::Cell;
    use std::rc::Rc;

    #[test]
    fn empty_dir_is_no_profile() {
        let dir = tempfile::tempdir().unwrap();
        assert!(!profile_exists(dir.path()));
        assert_eq!(
            has_saved_session(dir.path()).unwrap_err(),
            SavedSessionError::NoProfile
        );
    }

    #[test]
    fn marker_without_session_reads_false() {
        let dir = tempfile::tempdir().unwrap();
        // A valid but sessionless cookies.sqlite: present profile, no login.
        let conn = rusqlite::Connection::open(dir.path().join(PROFILE_MARKER)).unwrap();
        conn.execute_batch(
            "CREATE TABLE moz_cookies (name TEXT, value TEXT, host TEXT, path TEXT, \
             expiry INTEGER, isSecure INTEGER, isHttpOnly INTEGER, originAttributes TEXT)",
        )
        .unwrap();
        conn.execute(
            "INSERT INTO moz_cookies VALUES ('ct0', 'g', '.x.com', '/', 1786000000, 0, 0, '')",
            [],
        )
        .unwrap();
        conn.close().unwrap();
        assert!(profile_exists(dir.path()));
        assert!(!has_saved_session(dir.path()).unwrap());
    }

    #[test]
    fn wait_returns_on_first_session_sight() {
        let calls = Rc::new(Cell::new(0));
        let c2 = calls.clone();
        let r = wait_for_session(
            move || {
                c2.set(c2.get() + 1);
                c2.get() >= 3
            },
            || false,
            Duration::from_secs(300),
            Duration::from_secs(2),
            Instant::now,
            |_| {},
        );
        assert!(r.is_ok());
        assert_eq!(calls.get(), 3);
    }

    #[test]
    fn wait_reports_abort_and_timeout() {
        assert_eq!(
            wait_for_session(
                || false,
                || true,
                Duration::from_secs(300),
                Duration::from_secs(2),
                Instant::now,
                |_| {}
            ),
            Err(WaitError::Aborted)
        );
        let t0 = Instant::now();
        let now = Cell::new(t0);
        assert_eq!(
            wait_for_session(
                || false,
                || false,
                Duration::from_secs(10),
                Duration::from_secs(2),
                || now.get(),
                |d| now.set(now.get() + d),
            ),
            Err(WaitError::TimedOut)
        );
    }
}
